### DISCUSSION TITLE
Feedback/discussion for #336: API diffs

### DISCUSSION THREAD

**Discussion by rollacaster** [cljdoc/cljdoc#480] [Ideas]
Discussions about allowing to compare different libraries with different versions on cljdoc

**Comment by rollacaster** [comment]
Topic: Problem statement

**Reply by rollacaster** [reply]
Use cases we can solve
- Checking what has changed in a library between the version I'm on and the version that I'm upgrading to. [#1](https://github.com/cljdoc/cljdoc/issues/336#issue-473795876)
- See all breaking changes. [#1](https://github.com/cljdoc/cljdoc/issues/336#issuecomment-515962685)

**Comment by rollacaster** [comment]
Topic: How to find a breaking change

**Reply by rollacaster** [reply]
A list of possible breaking changes with sources we could show

- a removed namespaces/var [#1](https://github.com/borkdude/api-diff/blob/main/src/borkdude/api_diff.clj#L89)
- a removed arity[#1](https://github.com/cljdoc/cljdoc/issues/336#issuecomment-516077754) [#2](https://github.com/borkdude/api-diff/blob/main/src/borkdude/api_diff.clj#L84)
- a var became deprecated [#1](https://github.com/borkdude/api-diff/blob/main/src/borkdude/api_diff.clj#L86)
- a var became private [#1](https://github.com/borkdude/api-diff/blob/main/src/borkdude/api_diff.clj#L77)
- changed tests [#1](https://github.com/cljdoc/cljdoc/issues/336#issuecomment-515962685): If a function's tests have changed, that might indicate behavior has changed, or might not.

**Reply by rollacaster** [reply]
We might also use spec to find breaking/non-breaking changes. For example, if the new version updates the spec of a function but it still works with the old version (because only a new optional field was added) we would know it's a non-breaking change (or a breaking change if it fails)

(I am not sure if this is possible since I am quite new to spec)

**Reply by lread** [reply]
>     * changed tests [#1](https://github.com/cljdoc/cljdoc/issues/336#issuecomment-515962685): If a function's tests have changed, that might indicate behavior has changed, or might not.

This might be interesting info to report, but to me, does not fall under breaking change.

**Reply by lread** [reply]
This might be implied by "removed" or "became private" but we might also want to specifically list when a var/ns has become undocumented (via new presence of `:skip-wiki` or `:no-doc` metadata).

This could be argued as technically a non-breaking change (I mean you can still use undocumented thing!), but I think from the perspective of cljdoc which describes public APIs, it would be considered breaking.

**Reply by lread** [reply]
There is also a case I am exploring on the api-diff project.
What if we had `(defn thing [])` and this becomes `(def thing 42)`?
What once was callable is now not.
This kind of change is perhaps unlikely but worth noting if only to decide to scope it in or out.

**Comment by rollacaster** [comment]
Topic: Non-breaking changes

**Reply by rollacaster** [reply]
A list of possible non-breaking changes with sources we should show

- updated doc-strings [#1](https://github.com/cljdoc/cljdoc/issues/336#issuecomment-516077754)
- added namespaces/var's
- added functions

Cljdoc-specific changes:
- updated articles [#1](https://github.com/cljdoc/cljdoc/issues/336#issuecomment-529975413)

**Reply by lread** [reply]
There may also be added arities to an existing fn.

**Reply by lread** [reply]
Some of these things might be more interesting than others.
New vars and arities are, to me, more directly interesting than updated docstrings and articles.

I'm not saying that updated docstrings and articles are not interesting, but to me, they are much less interesting. If they were to be included in the UI, I'd rather my eye be drawn to the more interesting things first.

**Reply by martinklepsch** [reply]
Renamed function arguments might also be an interesting thing to highlight. They might hint at a breaking change but I think we should err on the side of not confusing the user by wrongly claiming something as a breaking change.

**Comment by rollacaster** [comment]
Topic: UI

**Reply by rollacaster** [reply]
Potential: Color Usage

If a function/var has been added, it would be marked with a green +
If a function/var has been removed, it would be marked with a red -
If a function/var has had changed, it would be marked with a blue ~ [#1](https://github.com/cljdoc/cljdoc/issues/336#issue-473795876)

**Reply by lread** [reply]
As a user, I'd be most interested in breakages. 
Then new opportunities.
Then maybe updated docs.

If we could design the UI somehow strongly accentuate the most interesting things first, I think that would help folks to see what is most important first.

**Reply by rollacaster** [reply]
I am not sure if breakages are most interesting for all cases. It could be nice if you can get a view that emphasizes on addition or breakages depending on your preference. 

Maybe similar to reversing directions in Google maps there could be a button to reverse if you compare the old to the new (focus on additions) or the new to the old (focus on breakage)

**Reply by lread** [reply]
For me, if I'm considering upgrading to a new version of a lib, the first thing I'm interested in is breakages. I want to know if my code will maybe break if I upgrade.
But, after that maybe I want to focus on non-breaking changes? Maybe. 
I guess you are trying to figure out what that would mean visually. De-emphasize breaking? Hide breaking? Emphasize non-breaking? 
All interesting UI elements to play with. 

The reversing directions/analogy idea is interesting. 
If we reverse the diff, are we answering the question "What happens if I downgrade?".

That's maybe useful, but I wouldn't want it to be the only way to learn about non-breaking changes, which is currently (soon to change) the only option for api-diff. It's too brain-twisty.

It might be helpful to start with a few basic use cases to focus the feature:
- As someone considering upgrading a library, I'd like to know if:
  - there are breaking changes (deletions)
  - there are soft breaking changes (deprecations, no-doc)
  - there is new functionality available (additions)
- As someone downgrading a library... 
  
And then you can decide which use cases to scope in and out for the first cut.
Folks will be very happy with a thoughtful minimal first cut.
They will give us feedback and ideas to incrementally improve the feature.

**Reply by martinklepsch** [reply]
My hunch is also we should prioritize the information like @lread suggested, first _breaking_, then _added_, then _modified_. 

A breaking change is MUST KNOW type of situation and usually there will be few of them so we will still have some screen real estate to surface the added and modified APIs. Also, people know how to scroll generally :)

**Comment by rollacaster** [comment]
Topic:  Make or Integrate

**Reply by rollacaster** [reply]
There are already libraries available which we could integrate into cljdoc, here are some pro/cons for doing so

### [api-diff](https://github.com/borkdude/api-diff)
**Pro**
- Actively developed
- Projects besides cljdoc benefit if we contribute there

**Con**
- cljdoc already has an analyzer so we would do the analysis twice
- Checks only for deprecation not additions
- Outputs only side effects (for now) and no data structure we could use

### [diff-apis](https://github.com/lread/diff-apis)
**Pro**
- cljdoc analyzer already in use. Analyze result can be passed to diff-apis which we probably (I did not check) already cache
- Includes some statistics for the diff which we could show on the UI as well
- Other projects interested in api-diff can use the library as well

**Con**
- Each call to `diff-apis` compares all the vars of the compared projects. This is not mandatory for cljdoc (but mandatory for calculating any statistics)

### Building directly into cljdoc's rendering
**Pro**
- No need to keep in sync with an external data format to describe diffs
- We already have the logic to traverse the API definitions and would not need to re-implement them
- We could only calculate the diffs which are currently visible and minimize resource usage

**Con**
- cljdoc could not benefit from upstream changes in the external library
- No benefits besides for cljdoc since the diff would not be reusable

**Reply by lread** [reply]
The joker who wrote diff-apis (it was me!) really doesn't want you to use it as anything but a place to get ideas. You don't want to use this code. It is not good. But you might be some good ideas from scanning the README.

Api-diff not only does its own analysis but it does static analysis.
There is nothing wrong with this (each approach has its pros and cons), but it is not the approach cljdoc has taken.
Cljdoc uses load time analysis via cljdoc-analyzer.
We could conceivably use api-diff if we could feed it data collected by cljdoc-analyzer.

I really like the api-diff code and am currently contributing to this project.
The code is simple and to the point.
At the very least you could consult it as a reasonable example of doing an API diff.

Whatever path we choose, I think separating the work of the diff outside the cljdoc rendering space makes good sense. This way the work of the diff is grokable/testable in absence of any rendering concerns.

**Comment by lread** [comment]
Nuance: Protocols

**Reply by lread** [reply]
When working on diff-apis I remember getting tripped up when comparing protocols.
This might have been partly due to the way cljdoc-analyzer collects data on protocols.
We might want to explore how protocols will be compared.

Api-diff does not currently treat protocols in any special way.
Maybe that would work.

But probably worth a bit of thought.

**Comment by lread** [comment]
Nuance: vars vs namespaces

**Reply by lread** [reply]
Api-diff triggered me to bring up this nuance. 
It currently reports only at the var level.

But for us, we probably also want to report at the namespace level?
If a namespace is added, we need to show each of its vars as new, do we do we want to indicate the namespace is entirely new? I'm guessing yes.

If a namespace has been removed, we need to show each of its vars as removed, do we want to indicate the namespace has been removed too? I'm guessing yes.

But worth a bit of thought probably.
